sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/AadOAuth2ClientAuthenticationJWKResolver.java

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.spring.cloud.autoconfigure.aad;
+
+import com.nimbusds.jose.jwk.JWK;
+import com.nimbusds.jose.jwk.RSAKey;
+import com.nimbusds.jose.util.Base64URL;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.core.ClientAuthenticationMethod;
+import org.springframework.util.Assert;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateEncodingException;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.security.interfaces.RSAPublicKey;
+import java.util.UUID;
+import java.util.function.Function;
+
+/**
+ * An {@link RSAKey} resolver function implementation parses the certificate locally.
+ *
+ * @since 4.3.0
+ * @see <a href="https://docs.microsoft.com/azure/active-directory/develop/active-directory-certificate-credentials">Certificate credentials</a>
+ */
+public class AadOAuth2ClientAuthenticationJWKResolver implements OAuth2ClientAuthenticationJWKResolver {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(AadOAuth2ClientAuthenticationJWKResolver.class);
+
+    private final String clientCertificatePath;
+    private final String clientCertificatePassword;
+
+    /**
+     * Creates a new instance of {@link AadOAuth2ClientAuthenticationJWKResolver}
+     * @param clientCertificatePath the client certificate path
+     * @param clientCertificatePassword the client certificate password
+     */
+    public AadOAuth2ClientAuthenticationJWKResolver(String clientCertificatePath,
+                                                    String clientCertificatePassword) {
+        this.clientCertificatePath = clientCertificatePath;
+        this.clientCertificatePassword = clientCertificatePassword;
+    }
+
+    @Override
+    public Function<ClientRegistration, JWK> resolve() {
+        return (clientRegistration) -> {
+            if (clientRegistration.getClientAuthenticationMethod().equals(ClientAuthenticationMethod.PRIVATE_KEY_JWT)) {
+                Assert.notNull(clientCertificatePath, "clientCertificatePath cannot be null");
+                Assert.notNull(clientCertificatePassword, "clientCertificatePassword cannot be null");
```

Comment:
This resolver is injected for each grant type provider, it will only be checked if the AOuth2 client has configured the authentication method `PRIVATE_KEY_JWT`.

Reply:
I am confused, how about we inject the resolver depending on whether a certificate is configured?